[12.0][IMP] hr_holidays_public: Create global calendar.event

ivantodorovich opened the pull request:
Notice there are 2 commits:

- The first one is just making the code a little bit more readable. No big changes there, but many lines are affected.

- The second one is the actual IMP

With this PR, public holidays will create events on the global calendar.
This is aligned with Odoo's creation of calendar events for hr.leaves, when validated.

Note that:
- The name of the event will be the name of the holiday.
- There's a "Holidays" tag applied to the event.
- The event will appear under the global calendar (Everybody's calendars)
- Since the event is created by the system, the user responsible is 'OdooBot' (System user)
- The event will be updated if the public holiday is updated
- The event will be deleted if the public holiday is deleted


![image](https://user-images.githubusercontent.com/1914185/62135799-259edf80-b2e3-11e9-8cc1-881e09255d33.png)

Maybe a known issue would be that users can edit the calendar event and the changes will not be blocked nor synced to the public holiday. I didn't want to get into this because that's the same behaviour we have with the events created for hr.leaves.

ivantodorovich committed:
[IMP] hr_holidays_public: Create global calendar.event

pedrobaeza commented on hr_holidays_public/models/hr_holidays_public.py:
```diff
@@ -135,26 +140,30 @@ class HrHolidaysPublicLine(models.Model):
         'Name',
         required=True,
     )
+
```
Please don't add useless empty lines. It's discouraged by PEP8 and doesn't increase readability. Indentation does that.

ivantodorovich replied:
Noted!

pedrobaeza commented on hr_holidays_public/models/hr_holidays_public.py:
```diff
@@ -57,8 +62,7 @@ def _compute_display_name(self):
                     line.year,
-                    line.country_id.name
-                )
+                    line.country_id.name)
```
Don't change ending characters deindented in other line. This is for readability, as you are doing for example in https://github.com/OCA/hr/pull/649/commits/a23796e805c8b65e228c90190bf43532263c1a69#diff-3080bf617f93d891bde828f3edd165d3R95

ivantodorovich replied:
Yeap, don't even know why I did that

pedrobaeza commented on hr_holidays_public/models/hr_leave_type.py:
```diff
@@ -10,9 +11,7 @@ class HrLeaveType(models.Model):
 
     exclude_public_holidays = fields.Boolean(
         string='Exclude Public Holidays',
+        help='If enabled, public holidays are skipped in leave days'
```
No added value

ivantodorovich replied:
Reverted.

alexey-pelykh commented:
@ivantodorovich please rebase on latest 12.0, there was a fix-merge recently, and please tidy up the commit history

alexey-pelykh reviewed (commented):
please rebase on latest 12.0, there was a fix-merge recently, and please tidy up the commit history

alexey-pelykh commented:
> Maybe a known issue would be that users can edit the calendar event and the changes will not be blocked nor synced to the public holiday. I didn't want to get into this because that's the same behaviour we have with the events created for hr.leaves.

@ivantodorovich can we block editing of such events entirely? If event is linked to a holiday, limit its `write` method and show error. But I guess it should be done via separate PR if anyone wishes to

pedrobaeza commented:
@ivantodorovich thanks for this. For me, taking Odoo's example of letting users to edit the generated entries is not a good reason for doing the same here, as we are in control of all elements. What you can do is:

* Add `public_holiday_id` many2one at calendar event level.
* Fill it in these entries.
* Overwrite `write` and `unlink` methods for triggering an exception when the m2o is filled unless an specific context is passed.
* When removing public holidays, call `unlink` of the calendar event with this context.
* The same when changing the public holidays.

What do you think?

Another doubt I have. What happen with regional public holidays? Are they put in the calendar as global?

ivantodorovich commented:
> please rebase on latest 12.0, there was a fix-merge recently, and please tidy up the commit history

Done!

I'm not sure we should block writing on it, maybe the user wants to change things like tags, descriptions, or even custom fields.. and that should be ok. So in that case a constraint restricting specific fields like date and name makes more sense to me.

Anyhow I honestly don't have much time these days, this was requested and financed by a customer as it is. I've added a note in the ROADMAP / Known issues file, so that we don't lose track of this, and we can implement that in a another PR later.

@pedrobaeza I've also added the note regarding regional public holidays.

> There are no restrictions to block users from modifying or removing calendar
events linked to public holidays. There's a suggestion to overload `write` and
`unlink` methods of `calendar.event`, but it might have other impacts like
users not being able to edit event tags, or even custom fields.

> Regional public holidays are shown in the public calendar. The regions will be
noted in the description of the event, but it'll be shown to all users. It'd
be good to have it show only for users in these regions.

pedrobaeza commented:
OK, I understand if you are not able to add that withing budget. A note in roadmap is enough for now.

alexey-pelykh commented:
I'm looking into #646 right now, it seems that if to implement creating global leave for each public holiday, events are going to be there in any case?

ivantodorovich commented:
@alexey-pelykh ~~yeap, they will be generated by the `hr.leave` model.~~
Let's discuss that in #646

ivantodorovich commented:
Actually, just realized that @jarroyomorales is talking about the `resource.calendar.leaves` model, that won't generate any event in the calendar.

jarroyomorales commented on hr_holidays_public/models/hr_holidays_public.py:
```diff
@@ -190,3 +193,43 @@ def _check_date_state_one(self):
+
+    @api.multi
+    def unlink(self):
+        self.mapped('meeting_id').unlink()
```
Here the `return super().unlink()` is missing

jarroyomorales commented:
@alexey-pelykh @ivantodorovich I think both things make sense and are compatible. On one hand we have the public holidays generating `resource.calendar.leaves` to solve some bugs mentioned in #646 and on the other hand we have this PR which looks absolutely fine for me since we would be doing the same than Odoo: Creating the leave -> Creating the `resource.calendar.leave` -> Creating the event.

ivantodorovich commented:
@jarroyomorales 
> Here the return super().unlink() is missing

Thanks!!!! Just sent the fixup

pedrobaeza reviewed (changes requested):
Please rebase squashing your fixup commit and removing that ugly merge commit that will scramble future migrations.

I see also some code that seems a bugfix for another feature (the one in hr_leave). Please put that in another PR.

ivantodorovich commented:
> Please rebase squashing your fixup commit and removing that ugly merge commit that will scramble future migrations.
> 
> I see also some code that seems a bugfix for another feature (the one in hr_leave). Please put that in another PR.

Upps.. sorry about that, it's done!

pedrobaeza reviewed (approved):
I consider fair to add Druidoo as co-author given the kind of contribution.

pedrobaeza commented:
/ocabot merge

pedrobaeza commented:
Can you please migrate this to v13 as well? Thanks!

ivantodorovich commented:
> Can you please migrate this to v13 as well? Thanks!

Thanks @pedrobaeza ! :)
13.0 PR: https://github.com/OCA/hr-holidays/pull/2
